### ISSUE TITLE
[bug] tools.cmake.cmaketoolchain:user_toolchain set by build_requires not consistently seen

### ISSUE THREAD

**Issue by puetzk** [conan-io/conan#9571]
### Environment Details (include every applicable attribute)
  * Operating System+version: Ubuntu 20.04
  * Compiler+version: 9.3.0-17ubuntu1~20.04
  * Conan version: 1.40
  * Python version: 3.8.10

### Steps to reproduce (Include if Applicable)

I have a package (represending a Yocto cross-compiler, further wrapped by winegcc) wants to inject custom embedded toolchain's CMAKE_TOOLCHAIN_FILE

```python
def package_info(self):
    ...
    self.conf_info["tools.cmake.cmaketoolchain:user_toolchain"] = os.path.join(self.package_folder,"EmbeddedToolchain.cmake")
```

Consumers (or really a profiles) would build_require this package target the system described by those settings.

This works if the consumer uses

```python
from conan.tools.cmake import CMakeToolchain
...
def generate(self):
     CMakeToolchain(self).generate()
```

but not if it uses 

```python
     generators = "CMakeToolchain"
```
the latter does not get the `include(/path/to/EmbeddedTollchain.cmake)` line in conan_toolchain.cmake (the user_toolchain block is missing).

I think the issue is that `_receive_conf` called (to integrate the info from build_requires)  during `write_toolchain`, which runs after the self._generator_manager.write_generators, so generates which run via the `generators =` attribute will not see the effects of any build_requires' conf_info, wheras the same generator invoked from generate() **will** have these values. This seems like an undesirable inconsistency.

https://github.com/conan-io/conan/blob/ff7b8e6703e8407773968517d68424b9ec59aa30/conans/client/generators/__init__.py#L229-L236
 
https://github.com/conan-io/conan/blob/ff7b8e6703e8407773968517d68424b9ec59aa30/conans/client/installer.py#L133-L137

**Comment by memsharded** [comment]
Thanks for the report, and the detailed investigation. Indeed a bug, fixed in https://github.com/conan-io/conan/pull/9573, and scheduled for next patch release 1.40.1.

Beware that this is quite experimental, there are quite a few unknowns, check the very simple implementation with a few TODOS:

```python
def _receive_conf(conanfile):
    """  collect conf_info from the immediate build_requires, aggregate it and injects/update
    current conf
    """
    # TODO: Open question 1: Only build_requires can define config?
    # TODO: Only direct build_requires?
    # TODO: Is really the best mechanism to define this info? Better than env-vars?
    # Conf only for first level build_requires
    for build_require in conanfile.dependencies.direct_build.values():
        if build_require.conf_info:
            conanfile.conf.compose(build_require.conf_info)
```

**Comment by puetzk** [comment]
> Beware that this is quite experimental, there are quite a few unknowns

Yeah :-). I just figured it was perhaps far enough along that I should try getting one of our cross-compiler/toolchain setups expressed using the new way and see what did/didn't work :-)

**Comment by memsharded** [comment]
Great, please try again in a few days, when we release 1.40.1, many thanks for trying the new features! :)

**Comment by memsharded** [comment]
Fixed by https://github.com/conan-io/conan/pull/9573, to be released in 1.40.1
